[GPU] Updated to get lockable memories for device inputs in prepare_input()

e-ddykim opened the pull request:
### Details:
 - Updated to get lockable memories for device inputs in prepare_input()
 - Without this PR, it fails to set f64 scalar tensor as input

### Tickets:
 - 98232, CVS-98722, CVS-98719 + set of conformance tickets like 98772

yeonbok commented:
Umm, I am curious why the usm device is used for input though it is not is_device_input. 
Could you elaborate more?

vladimir-paramuzov commented on src/plugins/intel_gpu/src/plugin/infer_request.cpp:
```diff
@@ -235,7 +235,19 @@ void InferRequest::set_input(const std::string& name, const Blob::Ptr& data) {
         if (data->buffer() == nullptr)
             IE_THROW(NotAllocated) << str_input_not_allocated << " Input name: \'" << name << "\'";
-        _inputs[name] = data;
+        if (_inputs.find(name) != _inputs.end() && _inputs[name]->byteSize() == data->byteSize()) {
```
We need to save user tensor, so the expected behavior is `_inputs[name] = data;` and then enqueue copy from user blob (system memory) to USM device blob (w/o lock)

e-ddykim replied:
I reverted the change and updated to allocate lockable memory for `_deviceInputs`. This makes it to store user tensors and treat them the same as other data types in `prepare_inputs()`. Thank you.

vladimir-paramuzov commented on src/plugins/intel_gpu/src/plugin/infer_request.cpp:
```diff
@@ -235,7 +235,19 @@ void InferRequest::set_input(const std::string& name, const Blob::Ptr& data) {
+            auto src_lock = data->cbuffer();
+            auto src_ptr = src_lock.as<uint8_t*>();
+            auto tgt_lock = _inputs[name]->cbuffer();
+            auto tgt_ptr = tgt_lock.as<uint8_t*>();
```
[random spot] Need a test

e-ddykim replied:
These changes are removed.

vladimir-paramuzov replied:
still need a test

e-ddykim replied:
I added a test to set tensor of f64 scalar type as input, which fails without this patch. Thank you.

e-ddykim commented:
> Umm, I am curious why the usm device is used for input though it is not is_device_input. Could you elaborate more?

I reverted the change and updated to get a lockable memory. Thank you.

p-durandin reviewed (commented):
Please check the performance on some limited set of models to be sure that not impacted

vladimir-paramuzov commented on src/plugins/intel_gpu/src/plugin/infer_request.cpp:
```diff
@@ -918,7 +918,7 @@ void InferRequest::prepare_input(const cldnn::primitive_id& inputName, Blob::Ptr
         }
     } else if (input_layout.is_static() && !is_dev_input && can_use_usm) {
-        allocate_dev_mem_if_needed(_deviceInputs, inputBlob, inputName, input_layout);
+        allocate_dev_mem_if_needed(_deviceInputs, inputBlob, inputName, input_layout, true);
```
You enforced need_lockable_memory to true, which means that we may have copy from system memory to usm *host* memory which will be assigned to first kernel then which is inefficient.

e-ddykim replied:
Because in the case of unsupported precisions, data must be converted to supported precisions using a write lock as below, I updated to get lockable memory only for the cases. Thank you.

https://github.com/openvinotoolkit/openvino/blob/fac45d79e4d2433e090f2a7d185fa990810eb74c/src/plugins/intel_gpu/src/plugin/infer_request.cpp#L958-L970

yeonbok replied:
How about replace it by adding an reorder kernel ?

e-ddykim replied:
The purpose of `prepare_input()` is to make the input data accessible from GPU. So, the conversion logic is executed only when `is_dev_input` is false, and if the input data is located in a location that cannot be accessed from GPU, such as a host buffer, it is not possible to process the conversion logic by a reorder kernel on the GPU device.

yeonbok replied:
>>  if the input data is located in a location that cannot be accessed from GPU, such as a host buffer,

What I meant was, how about add a new first reorder which gets cl_mem and writes the output to usm_device ?

yeonbok replied:
@vladimir-paramuzov  I have an additional question, is there an essential issue that the gpu plugin can't support I16/F64/I64 inputs? If not, I am just curious that whether they can be converted just adding a reorder to the network. (Even thought not to be done right now, but want to know the ultimate reason if any. :)

vladimir-paramuzov replied:
@yeonbok We just didn't have support of such data types in cldnn historically, so we can extend graph and reorder kernels to support more data types indeed. I think it's pretty easy for integer types, but with fp64 we may face an issue that it requires cl_khr_fp64 extension support by a device, and as I can see it's not supported by iGPUs at least. So we still need a custom conversion code.

yeonbok replied:
@vladimir-paramuzov  Thank you for the explanation of the history. Could I ask one more thing? In this portion, the type conversion is done only when is_dev_input is false. If is_dev_input is true, how those precisions are supported?

vladimir-paramuzov replied:
@yeonbok good question.. Looks like a possible bug in the impl. I think we need to add tests for such cases and ensure that those work.

yeonbok replied:
@vladimir-paramuzov  Thanks for the response. Then, how about 
1) Add the reorder strategy in a backlog & do it in the near future 
2) Add the exception & additional test for that case  in this PR? 
 @e-ddykim  @vladimir-paramuzov

e-ddykim replied:
I'm not clear what problem we're trying to solve, whether it's one problem or multiple.
After finishing this PR, how about clarifying what's next? @vladimir-paramuzov @yeonbok

yeonbok replied:
Since the fix itself for the original issues is merged to release branch, so now, I was seeing this PR as a fix for the missing hole of the original flow. So, it would be great if the observed missing part can be checked with simple work so that we can prevent future error. But of course, that can be separately done according to your priority and the required task to be done for the suggestion. Let me create a new backlog conaining both suggestion.

yeonbok replied:
I'd like to add one more observation : There is an issue in remote device mem + unsupported precisions. 
 ./benchmark_app -d GPU.1 -m ./models/~~~.xml  -ip I16 

[GPU] shared buffer has smaller size (768) than specified layout (1536)

Anyway I will create it as an issue.

e-ddykim commented:
> Please check the performance on some limited set of models to be sure that not impacted

I tested with 79 models for fp16 and fp32, and found no performance difference. Thank you.

vladimir-paramuzov commented on src/plugins/intel_gpu/src/plugin/infer_request.cpp:
```diff
@@ -918,7 +918,10 @@ void InferRequest::prepare_input(const cldnn::primitive_id& inputName, Blob::Ptr
     } else if (input_layout.is_static() && !is_dev_input && can_use_usm) {
-        allocate_dev_mem_if_needed(_deviceInputs, inputBlob, inputName, input_layout);
+        bool need_lockable_mem = false;
+        if (prec == Precision::I16 || prec == Precision::U16 || prec == Precision::FP64 || prec == Precision::U64 || prec == Precision::U32)
```
nit: we'd better have a common function to check unsupported input precision to ensure that this condition is aligned with copy logic

vladimir-paramuzov replied:
I'd suggest to cherry-pick this commit as is to release branch and improve code for master, so we minimize risk to break something in release.

e-ddykim replied:
I created a new PR to release. https://github.com/openvinotoolkit/openvino/pull/14532

e-ddykim replied:
I added a function `conv_to_supported_prec` to convert original precision to supported precision. Thank you.

yeonbok commented:
Please rebase the branch.
